### PR TITLE
Feature/gfsv16intocoupled

### PR DESCRIPTION
**Pull request by jkuang-exploro**
Script update.

Additional code and logics blocks from gfsv16 are included in the modularized scripts.

Next step is to add them into the namelists

### CODE REVIEW

File: jobs/JGLOBAL_FORECAST
Comment (JessicaMeixner-NOAA) [inline jobs/JGLOBAL_FORECAST:70]: Should this have if statements as in the develop branch: https://github.com/NOAA-EMC/global-workflow/blob/develop/jobs/JGLOBAL_FORECAST#L73-L79

File: ush/forecast_postdet.sh
Comment (JessicaMeixner-NOAA) [inline ush/forecast_postdet.sh]: why was DATA removed?

File: ush/forecast_postdet.sh
Comment (JessicaMeixner-NOAA) [inline ush/forecast_postdet.sh]: Why was DATA removed?

File: jobs/JGLOBAL_FORECAST
Comment (jkuang-exploro) [reply jobs/JGLOBAL_FORECAST:70]: I think it should be OK either way. These variables already appear in JJOB so I'm not sure why they are here at all. But I included them just in case there's some other concerns that I'm not yet aware of. 

These scripts probably need a extensive clean up of excess variables, seems like.

File: ush/forecast_postdet.sh
Comment (jkuang-exploro) [reply ush/forecast_postdet.sh]: It should be preserved. Let me fix it.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA  I know that the forecast script has to be manually merged, however I'm confused by this last commit 45341b6 "update /driver; /parm and /ecflow folder to match gfsv16" shouldn't those changes be incorporated instead through 

git merge develop 

instead of a manual merge? 

Also- I thought this PR was going to be just the first part of changing the forecast script, then there would be a second update with namelist changes.   Then at some point the actual "git merge develop" and moving to use in-line post for the atm model?

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA /driver; /parm and /ecflow are top directories containing files that won't be called in a CROW-based workflow; thus we have never use them, nor will in the future. For this reason, I simply update them by fully matching develop branch, largely for documentation and code management purpose.

As you know some files may needs more attention than others; so my approaches differ.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA if they are not used by CROW, all the more reason for retain their proper history by updating those files through "git merge develop" instead of by hand.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA That's what I've done for those directories.

**Comment by JessicaMeixner-NOAA** [comment]
How can you maintain their git histories if you copied them from the other branch instead of doing "git merge"? 

> @JessicaMeixner-NOAA That's what I've done for those directories.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA Oh. I thought they were equivalent. I will start over.

**Comment by JessicaMeixner-NOAA** [comment]
Instead of restarting, can we just revert this last commit?

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA Yes. Will do.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA thanks for reverting the last commit.  I ran a test with your updates and the forecast job failed.  I think it might be a disk space or user error so I'm rerunning and will let you know.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA my early error seems to have gone away however, now all of the grib2 files for the atmosphere are empty.  I suspect this is because only the forecast script has been moved to the new com directory structure but the atmosphere post scripts have not been updated to use the new com directory structure.  It might not end up being possible to do all of this in separate parts after all.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA OK let me check.

**Comment by JessicaMeixner-NOAA** [comment]
From my tests: 
COM from this PR branch: /work/noaa/marine/jmeixner/p6pr194/COMROOT/test01/
COM from top of coupled-crow: /work/noaa/marine/jmeixner/p6dev/COMROOT/test01/
that I was comparing

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA I am ready to push another commit containing resolutions of "git merge develop". Do you want me to hold up that?

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA No - go for it.  As is this PR cannot be merged, hopefully the update you will make will resolve the issues I saw.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA I am running into an error in my test case and it is likely related to WW3:

                    *** WAVEWATCH III Multi-grid shell ***
               =================================================


EXTCDE MPI_ABORT, IEXIT=  2002

slurmstepd: error: *** STEP 959619.0 ON Orion-24-24 CANCELLED AT 2020-12-02T15:11:52 ***

EXTCDE MPI_ABORT, IEXIT=  2002


EXTCDE MPI_ABORT, IEXIT=  2002

**Comment by jkuang-exploro** [comment]
Above is the log. Here are full information of this case:

EXPDIR: /work/noaa/marine/Jian.Kuang/expdir/casetest
ROTDIR: /work/noaa/marine/Jian.Kuang/comroot/casetest
DATA dir for forecast job: /work/noaa/marine/Jian.Kuang/dataroot/casetest/2013040100/gfs/fcst.62259

Would you please take a look at them when possible? I will keep investigate on my side as well.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA  thanks for the info - I'll look into it.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA can you check in your local changes you have so far?   I'll need them to make a few more changes to the wave init and prep jobs & scripts to get them to work.  Otherwise, I'll just redo them and make a PR back to this branch when I'm done.  Let me know your preference.

**Comment by jkuang-exploro** [comment]
Of course. I will do it right now.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA I've pushed all my local changes, mostly in the /workflow directory, for the additional config.wave* files and JJOB of waveinit.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA I have fixed the first problem and now am having others that will have to be sorted out by syncing the configs from the waves from the develop branch system to the crow-configure system (+ update values to be appropriate for the benchmark configuration)  I'm putting my updates here: https://github.com/JessicaMeixner-NOAA/global-workflow/tree/coupled/devmerge and will make a pr when the wave parts are ready.   It'll probably take another day or so.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA OK I will test run your other PRs now.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA I have made updates to have the branch match more closely with the develop branch (mostly from files that did not get deleted for some reason with the merge) and other updates related to the waves in my branch here: https://github.com/JessicaMeixner-NOAA/global-workflow/tree/coupled/devmerge  The ATM post still does not work due to the com restructuring I believe.     If PR #199 gets merged first, I can merge that in here before making a PR with my changes to your branch feature/gfsv16intocoupled if you want.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA OK. The test case finished just now for PR199. Let me merge it.

**Comment by jkuang-exploro** [comment]
Please review this PR once again. 

Test routine is identical with previous practice. Inline post is now a default setting, which user could turn off in the case file. Workflow layout will now change based on using inline post (default) or not.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA I will start my review of this PR... before I do since #210 is also ready to go should we merge that in first to feature/coupled-crow and pull those updates back in?  The only issue with that PR not being merged already is that it wasn't reviewed w/someone who has write permissions. 

Also another question: This completes the merge of everything from develop into feature/coupled-crow including the use of in-line post.  Does this also update the namelist changes that needed to be merged by hand into input.nml?

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA OK I will review PR#210. And yes this PR contains namelists updates.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA  maybe this is a one-off weird issue, but I just had this happen where the verify job is triggered very early somehow: 

```
       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
201304010000     ics.jgfs_emc_mom6ic                      985328           SUCCEEDED                   0         1          40.0
201304010000       gfs.prep.waveinit                      985330           SUCCEEDED                   0         1          19.0
201304010000       gfs.prep.waveprep                      985429           SUCCEEDED                   0         1           3.0
201304010000    gfs.forecast.highres                      985529              QUEUED                   -         0           0.0
201304010000       gfs.ocnpost.p_000                           -                   -                   -         -             -
201304010000       gfs.ocnpost.p_024                           -                   -                   -         -             -
201304010000            gfs.wavepost                           -                   -                   -         -             -
201304010000         gfs.wavepostpnt                           -                   -                   -         -             -
201304010000       gfs.jgfs_emc_vrfy                      985329           SUCCEEDED                   0         1           4.0
201304010000    archive.jgfs_archive                           -                   -                   -         -             -
201304010000                   final                           -                   -                   -         -             -
```

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA Yeah I am aware of that. Now drafting a fix.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA it didn't happen in my feature/coupled-crow branch test but I didn't see any changes from this branch that seemed like it was causing that so ?  Let me know if I should restart my test when the new code is ready.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA thanks for the fix. I'm starting a new test with the up to date code now.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA I fixed the triggering issue for post jobs. And I set inline_post to be default off. The test on Orion is running well so far. Would you please test it on Hera?

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA that's great news.  After PR #218 gets merged into feature/coupled-crow and is merged into this branch I will run a test on hera.  Since that PR is ready to be merged and we'd have to re-do the test after merging that in here, I think it's better to do that first and skip a test.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA thanks for merging in #218 and adding the fix by adding the variables to the defaults!  I noticed some additional commits on this branch for cleanup.  Are you ready for me to test this on hera or should I wait?

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA This PR is good to go per my test on Orion. Please test it on Hera and let me know how it goes. Thanks.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA, I will start my test on hera now.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA the run on hera failed the same way my run on orion failed.  See the log file 

/work/noaa/marine/jmeixner/p6pr194update/test01/COMROOT/test01/logs/2013040100/gfs.forecast.highres.log 

Important parts: 

```
0 + source /work/noaa/marine/jmeixner/p6pr194update/global-workflow/scripts/../ush/parsing_model_configure_FV3.sh
/work/noaa/marine/jmeixner/p6pr194update/global-workflow/scripts/../ush/parsing_model_configure_FV3.sh: line 76: syntax error near unexpected token `}'
/work/noaa/marine/jmeixner/p6pr194update/global-workflow/scripts/../ush/parsing_model_configure_FV3.sh: line 76: `}'
```

```
forrtl: severe (29): file not found, unit 101, file /work/noaa/marine/jmeixner/p6pr194update/test01/DATA/test01/2013040100/gfs/fcst.378149/model_configure

```

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA This is likely a typo. It is so strange that my test didn't capture it. Let me check.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA sorry I can't do this through the normal review as there are so many files with changes it's hard to go through and mark them on the file.  But I just checked the diffs in the input.nml creation from the namelist options that are in ush/parsing_namelists_FV3.sh and we need to update the following to match the develop branch: 

These are the variables in the develop branch options that I found that did not match the defaults in this PR: 
adjust_dry_mass = ${adjust_dry_mass:-".true."} 
lheatstrg    = ${lheatstrg-".false."}
lgfdlmprad   = ${lgfdlmprad:-".false."}
lgfdlmprad   = ${lgfdlmprad:-".true."}
effr_in      = ${effr_in:-".false."}
do_ugwp      = ${do_ugwp:-".true."}  
do_tofd      = ${do_tofd:-".true."}
reiflag = ${reiflag:-"2"}

In the develop branch cplwav needs to default to false:   
cplwav       = ${cplwav:-".false."} 
not just be: 
cplwav       = ${cplwav}

The if  $CCPP_SUITE blocks starting on line 201 could be updated to just be one if block following the "if [ $RUN_CCPP = "YES" ]; then" if block in the develop branch starting on line 1092 

lines 234-5 should be removed: 
  iovr_lw      = ${iovr_lw:-"3"}
  iovr_sw      = ${iovr_sw:-"3"}

Line 486: 
&nam_sfcperts 
should instead be   
$nam_sfcperts_nml

**Comment by JessicaMeixner-NOAA** [comment]
> @JessicaMeixner-NOAA This is likely a typo. It is so strange that my test didn't capture it. Let me check.

Are there local changes that did not get pushed?

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA Also doing a difference between the develop branch and this PR I saw that 
Files develop/sorc/build_ncep_post.sh and pr194/sorc/build_ncep_post.sh differ
which I don't think should necessarily be different. 

There were a couple of other files such as : 
Files develop/parm/config/config.gldas and pr194/parm/config/config.gldas differ
Files develop/env/HERA.env and pr194/env/HERA.env differ

which I believe is just there from recent commits to the develop branch.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA I fixed the model_configure script issue which is a simple syntax error. I will fix those name list variables that you mentioned, together with it.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA I think the RUN_CCPP block in the develop branch is not so clear because there are multiple CCPP suites and each of them may come with some additional name list variables. In other words, CCPP should be a selectable set instead of a  "true/false" variable. Indeed the RUN_CCPP block in the develop only takes care of the "FV3_GSD_v0" suite. Please correct me if I am wrong.

**Comment by jkuang-exploro** [comment]
> @JianKuang-NOAA sorry I can't do this through the normal review as there are so many files with changes it's hard to go through and mark them on the file. But I just checked the diffs in the input.nml creation from the namelist options that are in ush/parsing_namelists_FV3.sh and we need to update the following to match the develop branch:
> 
> These are the variables in the develop branch options that I found that did not match the defaults in this PR:
> adjust_dry_mass = ${adjust_dry_mass:-".true."} 
> lheatstrg    = ${lheatstrg-".false."}
> lgfdlmprad   = ${lgfdlmprad:-".false."}
> lgfdlmprad   = ${lgfdlmprad:-".true."}
> effr_in      = ${effr_in:-".false."}
> do_ugwp      = ${do_ugwp:-".true."} 
> do_tofd      = ${do_tofd:-".true."}
> reiflag = ${reiflag:-"2"}
> 
> In the develop branch cplwav needs to default to false:   
> cplwav       = ${cplwav:-".false."}
> not just be:
> cplwav       = ${cplwav}
> 
> The if $CCPP_SUITE blocks starting on line 201 could be updated to just be one if block following the "if [ $RUN_CCPP = "YES" ]; then" if block in the develop branch starting on line 1092
> 
> lines 234-5 should be removed:
> iovr_lw = ${iovr_lw:-"3"}
> iovr_sw = ${iovr_sw:-"3"}
> 
> Line 486: 
> &nam_sfcperts
> should instead be   
> $nam_sfcperts_nml

Line 486 "nam_sfcperts_nml", is identical with the develop branch. I don't know what should I do?

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA It's my understanding that we need these for all CCPP suites, not just "FV3_GSD_v0"  In the develop branch I think the suite that was recently tested was FV3_GFS_v16beta and we're going to be using the coupled version of that for prototype 6.

**Comment by jkuang-exploro** [comment]
> @JianKuang-NOAA It's my understanding that we need these for all CCPP suites, not just "FV3_GSD_v0" In the develop branch I think the suite that was recently tested was FV3_GFS_v16beta and we're going to be using the coupled version of that for prototype 6.

Yes. That's the reason why I think we shouldn't define CCPP as a YES/NO option. Instead we need to enable users to select suite by it's name.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA those options are needed for any CCPP suite at this time.  If you want to keep it for a specific suite file, please add an if block for "FV3_GFS_v16beta_coupled" with all of the options in the develop branch in that if block. 

You are correct on the nam_sfcperts_nml, it is already how it is in the develop branch.  I am not sure what I saw before.

File: workflow/cases/coupled_free_forecast.yaml
Comment (JessicaMeixner-NOAA) [inline workflow/cases/coupled_free_forecast.yaml]: @JianKuang-NOAA we aren't quite ready to jump to this yet.  can you please leave this as     
CCPP_SUITE: FV3_GFS_v15p2_coupled		
for the time being?

File: workflow/cases/prototype_5.1.yaml
Comment (JessicaMeixner-NOAA) [inline workflow/cases/prototype_5.1.yaml]: This isn't needed at this time.  We are not running a prototype 5.1 yet.  The next one will be p6.  We'll make a case file for it when we are ready, but we are not there yet.

File: workflow/cases/coupled_free_forecast.yaml
Comment (jkuang-exploro) [reply workflow/cases/coupled_free_forecast.yaml]: OK. I will revert this change.

**Comment by JessicaMeixner-NOAA** [comment]
I pulled the recent changes (up to 14a85cd3b7616cac213857b783bc44bce19b7d23) and tried again and the forecast still fails this time with: 
forrtl: severe (19): invalid reference to variable in NAMELIST input, unit -5, file Internal Formatted NML Read, line -1, position 16

Let me know when you are ready for me to run a test again.

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA It turns out that's because I included iovr_* variables in the "general" name list(for every CCPP suite). I fixed this issue. Please take another try on your end. I apologize for this mistake.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA the iovr and other namelist options look good.  Thanks for adding the CCPP suite options for  FV3_GFS_v16beta_coupled as well! 

Is the forecast model working again?  Is this ready for me to test again on hera?

**Comment by jkuang-exploro** [comment]
@JessicaMeixner-NOAA Yes it is working on my side again.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA my test on orion also succeeded. I started with a fresh clone and build so it's taking a little longer on hera.

@LydiaStefanova-NOAA and @jiandewang could you help check output Lydia note that atm files are now in the "atmos" subfolder
The test on orion output is here: /work/noaa/marine/jmeixner/p6pr194update/test02/COMROOT/test02/gfs.20130401/00 I'll post the hera test as soon as it's available.

**Comment by JessicaMeixner-NOAA** [comment]
@JianKuang-NOAA if we wanted to get bit-reproducible answers, we could by overwriting the default values.. however since we're in the process of upgrading, perhaps as long as things are running this is not an issue.  We do want to make sure that everything is working as expected though.

**Comment by JessicaMeixner-NOAA** [comment]
@jiandewang @LydiaStefanova-NOAA the hera test output is here: 
/scratch2/NCEPDEV/climate/Jessica.Meixner/p6pr194update/test01/COMROOT/test01/gfs.20130401/00 

The hera test ran to "succeed" although the archive job is still running.   The archive job is not completely correct as @jiandewang pointed out earlier but we can address that in a later PR.

**Comment by JessicaMeixner-NOAA** [comment]
Copying @LydiaStefanova-NOAA's comment's to the PR from the issue: 
In atmos/ the grib2 files look ok to me.

In wave/ the gfswave.t00z.global.0p50.fXXX.grib2 files all fields I checked are constant (i.e. identical value at any lat/lon)

Two other things I noticed, that may or may not be relevant here, mentioning them in case they are:
(a) The variable called "la_mod" in the waveocn_*nc files has possibly strange values (e.g the max is 5.37e+34 in wavocn_2013_04_01_03.nc, 8.9301E+08 in wavocn_2013_04_02_21.nc). I haven't looked at this type of files before, so I don't know if this is normal.
(b) the archlist contents (now under atmos/archlist) are different from what we had for P5. I can describe the differences in more detail if this is relevant now.

And from @jiandewang: 
checking ocean outputs show reasonable value (both orion and hera), but confirmed with Lydia's point about LA_MOD

And from me: 
 We know that the arch list is not perfect and will take care of that later. The bad wave grib files is another known issue that this PR is not supposed to address. the la_mod does have very large values at times. That was there before. 

Thanks to @jiandewang and @LydiaStefanova-NOAA for helping check things.

**Review by JessicaMeixner-NOAA — APPROVED** [review]
Thanks @JianKuang-NOAA

**Pull request merged**
